File: app/src/main/java/priv/rtbishop/coopstat/ui/StreamFragment.java
package priv.rtbishop.coopstat.ui;

import android.annotation.SuppressLint;
import android.content.Context;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.ViewTreeObserver;
import android.webkit.WebView;
import android.webkit.WebViewClient;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;

import java.util.Objects;

import priv.rtbishop.coopstat.R;
import priv.rtbishop.coopstat.vm.MainViewModel;

public class StreamFragment extends Fragment {

    private MainViewModel mViewModel;
    private WebView mStreamView;
    private Context mContext;
    private int mScale;

    @Override
    public void onAttach(@NonNull Context context) {
        super.onAttach(context);
        mContext = context;
    }

    public View onCreateView(@NonNull LayoutInflater inflater,
                             ViewGroup container, Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_stream, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        mViewModel = new ViewModelProvider(Objects.requireNonNull(getActivity())).get(MainViewModel.class);
        setupViews(view);

        if (mViewModel.isConnected()) {
            loadUrl();
        } else {
            Toast.makeText(mContext, R.string.connection_not_found, Toast.LENGTH_LONG).show();
        }
    }

    private void setupViews(@NonNull View view) {
        mStreamView = view.findViewById(R.id.web_view_stream);

        ViewTreeObserver viewTreeObserver = mStreamView.getViewTreeObserver();
        if (viewTreeObserver.isAlive()) {
            viewTreeObserver.addOnGlobalLayoutListener(new ViewTreeObserver.OnGlobalLayoutListener() {
                @Override
                public void onGlobalLayout() {
                    mStreamView.getViewTreeObserver().removeOnGlobalLayoutListener(this);
                    mScale = (int) ((mStreamView.getWidth() / 1300f) * 100);
                }
            });
        }
    }

    private void loadUrl() {
        mStreamView.post(new Runnable() {
            @SuppressLint("SetJavaScriptEnabled")
            @Override
            public void run() {
                mStreamView.getSettings().setJavaScriptEnabled(true);
                mStreamView.setWebViewClient(new WebViewClient());
                mStreamView.setInitialScale(mScale);
                mStreamView.loadUrl(mViewModel.getProxyUrl() + "/stream_simple.html");
            }
        });
    }
}

File: app/src/main/java/priv/rtbishop/coopstat/ui/StreamFragment.kt
package priv.rtbishop.coopstat.ui

import android.annotation.SuppressLint
import android.os.Bundle
import android.view.View
import android.view.ViewTreeObserver
import android.webkit.WebView
import android.webkit.WebViewClient
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProvider
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import priv.rtbishop.coopstat.R
import priv.rtbishop.coopstat.vm.MainViewModel

class StreamFragment : Fragment(R.layout.fragment_stream) {

    private lateinit var swipeLayout: SwipeRefreshLayout
    private lateinit var viewModel: MainViewModel
    private lateinit var streamView: WebView
    private lateinit var webViewClient: WebViewClient
    private var scale: Int = 0

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        viewModel = ViewModelProvider(activity as MainActivity).get(MainViewModel::class.java)
        webViewClient = WebViewClient()

        swipeLayout = view.findViewById(R.id.swipe_layout)
        streamView = view.findViewById(R.id.stream_view)
        val observer = streamView.viewTreeObserver
        if (observer.isAlive) {
            observer.addOnGlobalLayoutListener(object : ViewTreeObserver.OnGlobalLayoutListener {
                override fun onGlobalLayout() {
                    streamView.viewTreeObserver.removeOnGlobalLayoutListener(this)
                    scale = (streamView.width / 1300f * 100).toInt()
                }
            })
        }

        val proxyUrl = viewModel.proxyUrl.value
        if (proxyUrl.isNullOrEmpty()) {
            swipeLayout.isRefreshing = true
            viewModel.setupProxyConnection()
        } else {
            loadUrl(proxyUrl)
        }

        viewModel.proxyUrl.observe(viewLifecycleOwner, {
            swipeLayout.isRefreshing = false
            loadUrl(it)
        })

        swipeLayout.setOnRefreshListener { viewModel.setupProxyConnection() }
    }

    @SuppressLint("SetJavaScriptEnabled")
    private fun loadUrl(url: String) {
        streamView.post {
            streamView.settings.javaScriptEnabled = true
            streamView.webViewClient = webViewClient
            streamView.setInitialScale(scale)
            streamView.loadUrl("$url/stream_simple.html")
        }
    }
}